DiverseSelector/methods/partition.py

```diff
@@ -181,168 +172,371 @@ def select_from_cluster(self, X, size, cluster_ids=None):
+        This is done by doing a linear interpolation from integer indices and points, where
+        it is then evaluated on a uniform grid with number of bins as the spacing in each axis.
 
         Parameters
         ----------
-        arr: np.ndarray
-            Coordinate array of points
-        num_selected: int
-            Number of molecules that need to be selected.
-        cluster_ids: np.ndarray
-            Indices of molecules that form a cluster
+        X: ndarray of shape (n_samples, n_features)
+           Feature matrix of `n_samples` samples in `n_features` dimensional space.
+        nbins_axis: int
+            Number of bins along each axis or feature dimension.
 
         Returns
         -------
-        selected: list
-            List of ids of selected molecules
+        unique_bin_indices: ndarray(int,)
+            Unique (without duplication) bin indices that have at least one sample point.
+            These are integer tuples :math:`(i_1, \cdot, i_\text{n_features})` with elements
+            corresponding to the bin index along each axis/dimension of feature space.
+            `inverse_ids` contains indices of `unique_bins_ids` for each of the :math:`N` points that
+            it is assigned to.
+        inverse_indices: ndarray(int,)
+            Indices of the unique bins (along specified axis) that can be used to reconstruct bin
+            index of each sample (`unique_bin_indices[inverse_indices]` gives bin index array).
         """
-        if cluster_ids is not None:
-            arr = arr[cluster_ids]
+        n_samples = len(X)
+        # to obtain the lower and upper range of bins so that each bin has equal number of points,
+        # we interpolate the feature value for indices delineating the range of bins.
+        # I.e., sorting the features values along one axis, the (xp, fp) pairs are formed where
+        # xp denote the integer index of `n_samples points and fp the corresponding feature values.
+        # now, the value of feature is interpolated for the indices corresponding to lower and
+        # upper ranges of bins denoted by x. The x indices are obtained by evenly dividing the
+        # (0, n_samples) into nbins_axis bins (hence nbins_axis + 1 for number of samples indices).
+        # Obviously, in this way, the bins have roughly `n_samples` points.
+        # Note that the starting and ending indices of the bins are always 0 and n_samples,
+        # corresponding to the minimum and maximum of the feature values.
+        # The resulting bins_edge defines a monotonically increasing array of bin edges, including
+        # the rightmost edge, allowing for non-uniform bin widths.
+        # The bins_edge[0] and bins_edge[-1] correspond to the min and max of the feature values.
+        bins_edge = np.interp(
+            x=np.linspace(0, n_samples, nbins_axis + 1), xp=np.arange(n_samples), fp=np.sort(X)
+        )
```

Comment:
@Ali-Tehrani, I have one last question. Shouldn't we use `x=np.linspace(0, n_samples - 1, nbins_axis + 1)`, so that the ending index corresponds to the index to the last sample point? Otherwise, we are extrapolating outside the range of feature values (for a given dimension).

Reply:
Extrapolation to the outside doesn't occur.  The numpy interpolate function [np.interp](https://numpy.org/doc/stable/reference/generated/numpy.interp.html) has two attributes called right/left, that if the value x is outside the interpolating domain, then it returns the closest point fp[-1]/fp[0], respectively.